### ISSUE TITLE
Gaussian Mixture Models for streaming data

### ISSUE THREAD

**Issue by vsoch** [online-ml/river#871]
Hi @MaxHalford! I want to introduce you to @harshithamenon, we were looking at different model choices for our errors project, and we didn't find anything for GMM. I found that "Mixture Models" are on your [roadmap](https://www.notion.so/d1e86fcdf21e4deda16eedab2b3361fb?v=503f44740b8b44a99a961aa96e9e46e1&p=a162fd78983f4fc8be36b2d5e9bc6ad2):

![image](https://user-images.githubusercontent.com/814322/157507126-80f1f16f-ad48-451c-bde0-db3bbb3a5fcd.png)

And also there seems to be an old issue that started some discussion? https://github.com/online-ml/river/issues/71

I think we like the [paper you had in mind](https://hal.archives-ouvertes.fr/hal-00201327/document) and we were brainstorming these might be of interest too:

 - [Clustering of Data Streams With Dynamic Gaussian Mixture Models: An IoT Application in Industrial Processes ](https://oa.upm.es/54569/1/INVE_MEM_2018_293573.pdf)
 - [Gradient-Based Training of Gaussian Mixture Models for High-Dimensional Streaming Data](https://link.springer.com/article/10.1007/s11063-021-10599-3)

I think we would be interested in giving this implementation a shot - @harshithamenon is really good at ML, and (well you know me) I'm the builder monkey :monkey: :laughing: There is no rush - we were thinking of chatting next week, but I wanted to open this issue to start up discussion about 1. if we could work on this, and 2. if yes, if we could talk through the approach to try first and design for the implementation, etc. Thank you!

**Comment by MaxHalford** [comment]
Hey there! Pleased to e-meet you @harshithamenon.

> There is no rush

Indeed there isn't, opensource should be fun :)

> I wanted to open this issue to start up discussion about 1. if we could work on this, and 2. if yes, if we could talk through the approach to try first and design for the implementation, etc. Thank you!

Good idea! I find that it helps to find a good dataset/usecase to help develop the model. I'd be more than happy to help with technical details, such as River conventions, working with dictionaries, evaluating models online, etc. Feel free to ask questions and I'll do what I can.

**Comment by MaxHalford** [comment]
@vsoch may I have your email address? I wanted to talk to you about a recent development regarding the deployment tool you're working on.

**Comment by vsoch** [comment]
Yep! Will send to you.

**Comment by MaxHalford** [comment]
I really like [this example](https://scikit-lego.readthedocs.io/en/latest/mixture-methods.html) from scikit-lego where they use mixture models for anomaly detection. A lot of people use River for anomaly detection, and there's not many methods that work online. So it's a very worthwhile direction to pursue.

**Comment by vsoch** [comment]
Would you want to use the [dataset directly from scikit-learn](https://github.com/scikit-learn/scikit-learn/blob/7e1e6d09bcc2eaeba98f7e737aac2ac782f0e5f1/sklearn/datasets/_samples_generator.py#L723) or try to add an implementation of it here?

**Comment by MaxHalford** [comment]
We have a [`stream.iter_sklearn_dataset`](https://riverml.xyz/latest/api/stream/iter-sklearn-dataset/) which avoids the need to copy/paste their datasets. It convert each observation into a dict and adds the correct feature names :)

**Comment by vsoch** [comment]
No worries! I don't know if @harshithamenon ever had bandwidth for this - @harshithamenon do you want to comment? I'd be happy to take it over if you started something and it's just a matter of preparing a PR.

**Comment by geoffroyO** [comment]
Hello, I would like to reopen this post maybe to work on it.

I coded a version of the online EM for the GMM of Cappé and Moulines entirely in with numpy to which I added

1. A training possibility with mini-batch
2. Polyak-averaging 
3. The version I made doesn't do any inversion, nor calculation of the covariance determinant for density evaluation or statistics: everything is updated online which is much faster than the original version.

However I would like to see how it would be possible to implement all this on river, for the moment my biggest problems are:

1. In practice we do a warm-up phase where only the statistics are updated so we don't learn the parameters during the first iterations: using the "learn_one" method would not be fair since we don't learn the parameters here but just the statistics.
2. Concerning the initialization I often wait to have some points to make a k-means, but how to implement that?

The idea I have to get around these "initialization" problems would be to make a "warm-up" method that would accumulate the first n points and then after a certain number of accumulated data we could finally use learn_one method....

**Comment by geoffroyO** [comment]
Hey @MaxHalford, 

Yep I'm using Sherman-Morisson formula alongs with a kind of Sherman-Morrisson for the determinant, details are in redaction !

Concerning the warm-up phase it often adds more stability but with a good initialization of the statistics it may not be necessary. Nevertheless I think that it will be necessary to wait to have some points to be able to initialize the parameters well if the user chooses the automatic initialization. I will take a look at the online k-means!

I'll try to make a first proposition next week if I find some time.
